msprime/tables.py

```diff
@@ -53,33 +54,86 @@ def text_decode_metadata(encoded):
     return base64.b64decode(encoded.encode('utf8'))
 
 
+NodeTableRow = collections.namedtuple(
+    "NodeTableRow",
+    ["flags", "time", "population", "metadata"])
+
+
+EdgeTableRow = collections.namedtuple(
+    "EdgeTableRow",
+    ["left", "right", "parent", "child"])
+
+
+MigrationTableRow = collections.namedtuple(
+    "MigrationTableRow",
+    ["left", "right", "node", "source", "dest", "time"])
+
+
+SiteTableRow = collections.namedtuple(
+    "SiteTableRow",
+    ["position", "ancestral_state", "metadata"])
+
+
+MutationTableRow = collections.namedtuple(
+    "MutationTableRow",
+    ["site", "node", "derived_state", "parent", "metadata"])
+
+
+ProvenanceTableRow = collections.namedtuple(
+    "ProvenanceTableRow",
+    ["timestamp", "record"])
+
+
 class NodeTable(_msprime.NodeTable):
     """
-    Class for tables describing all nodes in the tree sequence, of the form
-        id     is_sample  population   time
-        0      1          0            0.0
-        1      1          1            0.0
-        2      0          0            0.0
-        3      1          0            0.5
-        4      0          2            2.1
-    Node IDs are *not* recorded; rather the `id` column shows the row index, so
-    that the `k`-th row describes the node whose ID is `k`.  `is_sample`
-    records whether the node is a sample (=1) or not (=0).  `population` is an
-    integer population ID, and `time` is the time since that individual was
-    born, as a float.
-
-    Requirements:
-        1. All birth times must be greater than or equal to zero.
-
-    It is not required that the `time` column be ordered or that all samples
-    must be at the top.
+    A table defining the nodes in a tree sequence. See the
+    :ref:`definitions <sec-node-table-definition>` for details on the columns
+    in this table and the
+    :ref:`tree sequence requirements <sec-valid-tree-sequence-requirements>` section
+    for the properties needed for a node table to be a part of a valid tree sequence.
+
+    Example::
+
+        >>> t = msprime.NodeTable()
+        >>> t.add_row(flags=0, time=1)
+        0
+        >>> t.add_row(flags=1, time=2)
+        1
+        >>> print(t)
+        id      flags   population      time    metadata
+        0       0       -1      1.00000000000000
+        1       1       -1      2.00000000000000
+        >>> print(t[0])
+        NodeTableRow(flags=0, time=1.0, population=-1, metadata=b'')
+        >>> print(t[-1])
+        NodeTableRow(flags=1, time=2.0, population=-1, metadata=b'')
+        >>> len(t)
+        2
+
+    :warning: The numpy arrays returned by table attribute accesses are **copies**
+        of the underlying data. In particular, this means that you cannot edit
+        the values in the columns by updating the attribute arrays.
+
+        **NOTE:** this behaviour may change in future.
+
+    :ivar time: The array of time values.
+    :vartype time: numpy.ndarray, dtype=np.float64
```

Comment:
These and other docs below do a nice job laying out the dtypes. This info is duplicated in several places. 

An alternative is to put the type this info in code somewhere, then and referring? Something like below (slightly abusing the namedtuple) which could be a static member of `NodeTable`.

```py
 NodeTableTypes = NodeTableRow(flags=np.uint32, time=np.float64,
                                 population=np.int32, metadata=np.int8)
```

At the end of the day, though, keeping the class, row, and 2 column interfaces up to date is maybe less than figuring out and maintaining a code API to the type info

Reply:
Yeah, I was trying to figure out something like this. The short version is the user will want to have the dtype information on hand, and it's too much trouble trying to figure out an automatic way of doing it. The dtypes can't change very much, so hopefully it won't be a maintenance problem.

Reply:
Sounds good